File: mobile/src/services/api/game.ts
import { apiClient } from "./client";
import type { GameQuestion, SubjectId, ChildProgress } from "../../../shared/types";

export type { GameQuestion, ChildProgress };

// ---------------------------------------------------------------------------
// Difficulty advisor types
// ---------------------------------------------------------------------------

export interface ResponseRecord {
  question_text: string;   // what Zoey asked
  selected_text: string;   // answer the child tapped (text, not id)
  correct_text: string;    // the correct answer text
  is_correct: boolean;
}

export interface DifficultyAdjustRequest {
  child_id: string;
  child_name: string;
  child_age: number;
  subject_id: SubjectId;
  skill_id: string;              // = subject_id now; granular skill IDs are future work
  recent_responses: ResponseRecord[];  // exactly 3
  current_difficulty: number;
}

export interface DifficultyAdjustResponse {
  new_difficulty: number;
  zoey_cue: "encourage" | "challenge" | "scaffold";
  zoey_message: string;
  tts_url: string | null;
}

export interface QuestionRequest {
  child_id: string;
  subject_id: SubjectId;
  child_name: string;
  child_age: number;
  level: number;
}

export interface AnswerRequest {
  child_id: string;
  subject_id: SubjectId;
  question_id: string;
  selected_id: string;
  correct_id: string;
  scored: boolean;
  level: number;
  session_question_count: number;
  session_correct_count: number;
}

export interface AnswerResponse {
  is_correct: boolean;
  new_level: number | null;
  accuracy_rate: number;
}

export const fetchQuestion = async (req: QuestionRequest): Promise<GameQuestion> => {
  const { data } = await apiClient.post<GameQuestion>("/api/v1/game/question", req);
  return data;
};

export const submitAnswer = async (req: AnswerRequest): Promise<AnswerResponse> => {
  const { data } = await apiClient.post<AnswerResponse>("/api/v1/game/answer", req);
  return data;
};

export const fetchProgress = async (
  childId: string,
  subjectId: SubjectId,
): Promise<ChildProgress> => {
  const { data } = await apiClient.get<ChildProgress>(
    `/api/v1/game/progress/${childId}/${subjectId}`,
  );
  return data;
};

export const adjustDifficulty = async (
  req: DifficultyAdjustRequest,
): Promise<DifficultyAdjustResponse> => {
  const { data } = await apiClient.post<DifficultyAdjustResponse>(
    "/api/v1/game/adjust-difficulty",
    req,
  );
  return data;
};

File: backend/app/models/game.py
"""
Game engine models — questions, answers, progress.

These mirror the TypeScript types in shared/types/index.ts.
"""
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional
import uuid

from pydantic import BaseModel, Field


class GameSubjectId(str, Enum):
    reading = "reading"
    math = "math"
    science = "science"
    social_studies = "social_studies"
    sel = "sel"
    arts = "arts"
    health = "health"


class QuestionType(str, Enum):
    multiple_choice = "multiple_choice"


class AnswerChoice(BaseModel):
    id: str           # "a" | "b" | "c" | "d"
    text: str         # display text
    emoji: Optional[str] = None  # optional visual hint


class GeneratedQuestion(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    subject: GameSubjectId
    level: int        # 1–5
    type: QuestionType = QuestionType.multiple_choice
    prompt: str       # what Zoey reads aloud + displays
    choices: list[AnswerChoice]
    correct_id: str   # matches one AnswerChoice.id
    zoey_correct: str # Zoey's celebration (spoken after correct answer)
    zoey_wrong: str   # Zoey's encouragement (spoken after wrong answer)
    # False for SEL / Arts — no right/wrong, only reflective feedback
    scored: bool = True
    # TTS URLs filled in after synthesis
    tts_prompt_url: Optional[str] = None
    tts_correct_url: Optional[str] = None
    tts_wrong_url: Optional[str] = None


class QuestionRequest(BaseModel):
    child_id: str
    subject_id: GameSubjectId
    child_name: str
    child_age: int
    level: int = 1


class AnswerRequest(BaseModel):
    child_id: str
    subject_id: GameSubjectId
    question_id: str
    selected_id: str
    correct_id: str
    scored: bool = True
    level: int
    session_question_count: int  # total questions answered this session so far
    session_correct_count: int   # correct answers this session so far


class AnswerResponse(BaseModel):
    is_correct: bool
    new_level: Optional[int] = None   # set only when difficulty changes
    accuracy_rate: float              # lifetime accuracy for this subject


class ChildProgress(BaseModel):
    subject_id: GameSubjectId
    level: int = 1
    questions_answered: int = 0
    correct_count: int = 0
    accuracy_rate: float = 0.0
    last_played: Optional[datetime] = None


# ---------------------------------------------------------------------------
# Adaptive difficulty advisor — GPT-4o mid-session coaching
# ---------------------------------------------------------------------------

class ResponseRecord(BaseModel):
    """A single answer event, used as input to the difficulty advisor."""
    question_text: str   # the prompt Zoey read aloud
    selected_text: str   # what the child tapped (answer text, not id)
    correct_text: str    # the correct answer text
    is_correct: bool


class DifficultyAdjustRequest(BaseModel):
    child_id: str
    child_name: str
    child_age: int
    subject_id: GameSubjectId
    skill_id: str              # subject for now; future: "reading.phonics"
    recent_responses: list[ResponseRecord]   # exactly 3
    current_difficulty: int    # 1–5


class DifficultyAdjustResponse(BaseModel):
    new_difficulty: int                    # 1–5
    zoey_cue: str                          # "encourage" | "challenge" | "scaffold"
    zoey_message: str                      # Zoey speaks this (scaffold/challenge only)
    tts_url: Optional[str] = None
